controllers/disruption_controller.go

```diff
@@ -272,6 +334,132 @@ func (r *DisruptionReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 }
 
+// createInstanceSelectorCache creates this instance's cache if it doesn't exist and attaches it to the controller
+func (r *DisruptionReconciler) manageInstanceSelectorCache(instance *chaosv1beta1.Disruption) error {
+	// clean up potential expired cache contexts based on context error return
+	deletionList := []string{}
+
+	for key, contextTuple := range r.CacheContextStore {
+		if contextTuple.Ctx.Err() != nil {
+			deletionList = append(deletionList, key)
+		}
+	}
+
+	for _, key := range deletionList {
+		delete(r.CacheContextStore, key)
+	}
+
+	// remove check when StaticTargeting is defaulted to false
+	if instance.Spec.StaticTargeting == nil {
+		r.log.Debugw("StaticTargeting pointer is nil")
+	}
+
+	if instance.Spec.StaticTargeting == nil || *instance.Spec.StaticTargeting {
+		return nil
+	}
+
+	disNamespacedName := types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}
+	disSpecHash, err := instance.Spec.HashNoCount()
+
+	if err != nil {
+		return fmt.Errorf("error getting disruption hash")
+	}
+
+	disCacheHash := disNamespacedName.String() + disSpecHash
+	disCompleteSelector, err := targetselector.GetLabelSelectorFromInstance(instance)
+
+	if err != nil {
+		return fmt.Errorf("error getting instance selector: %w", err)
+	}
+	// if it doesn't exist, create the cache context to re-trigger the disruption
+	if _, ok := r.CacheContextStore[disCacheHash]; !ok {
+		// create the cache/watcher with its options
+		var cacheOptions k8scache.Options
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Node{}: {Label: disCompleteSelector},
+				},
+			}
+		} else {
+			cacheOptions = k8scache.Options{
+				SelectorsByObject: k8scache.SelectorsByObject{
+					&corev1.Pod{}: {Label: disCompleteSelector},
+				},
+				Namespace: instance.Namespace,
+			}
+		}
+
+		cache, err := k8scache.New(
+			ctrl.GetConfigOrDie(),
+			cacheOptions,
+		)
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		// attach handler to cache in order to monitor the cache activity
+		var info k8scache.Informer
+
+		if instance.Spec.Level == chaostypes.DisruptionLevelNode {
+			info, err = cache.GetInformer(context.Background(), &corev1.Node{})
+		} else {
+			info, err = cache.GetInformer(context.Background(), &corev1.Pod{})
+		}
+
+		if err != nil {
+			return fmt.Errorf("cache gen error: %w", err)
+		}
+
+		info.AddEventHandler(DisruptionSelectorHandler{disruption: instance, reconciler: r})
+
+		// start the cache with a cancelable context and duration, and attach it to the controller as a watch source
+		ch := make(chan error)
+		cacheCtx, cacheCancelFunc := context.WithTimeout(context.Background(), instance.Spec.Duration.Duration()+*r.ExpiredDisruptionGCDelay*2)
+
+		go func() { ch <- cache.Start(cacheCtx) }()
+
+		r.CacheContextStore[disCacheHash] = CtxTuple{cacheCtx, cacheCancelFunc}
```

Comment:
A context doesn't remember its own cancel func? that's disappointing, but this is a good solution